### PR TITLE
DDL: do not Reload() for 'CREATE TEMPORARY' and 'DROP TEMPORARY' statements

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

In `query_executor.go`, in `execDDL()`, we differentiate two cases:

1. The query is an Online DDL
2. The query is not Online DDL

In the latter case, we always invoke `Reload()` (specifically `ReloadAtEx()`) to reload table schema. 
However, for `CREATE TEMPORARY TABLE` and `DROP TEMPORARY TABLE` this is wasteful:

- Only the connections which created or dropped the table is at all aware of the existence of the table
- `Reload()` adds an overhead which becomes significant if temporary tables are created and dropped intensively.

With this PR we skip `Reload()` for temporary tables.

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 1]
```diff
@@ -657,20 +657,31 @@ func (qre *QueryExecutor) execDDL(conn *StatefulConnection) (*sqltypes.Result, e
...
-
+	isTemporaryTable := false
+	switch stmt := qre.plan.FullStmt.(type) {
+	case *sqlparser.CreateTable:
+		isTemporaryTable = stmt.IsTemporary()
+	case *sqlparser.DropTable:
+		isTemporaryTable = stmt.IsTemporary()
+	}
+
+	if !isTemporaryTable {
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/query_executor.go:668]: In the analyzeDDL plan
we mark in the plan `NeedsReservedConn` to true, so we can use that field here and ignore the `reload`

```go
return &Plan{PlanID: PlanDDL, FullQuery: fullQuery, FullStmt: stmt, NeedsReservedConn: stmt.IsTemporary()}, nil
```

**Review by harshit-gangal — COMMENTED** [review]
Can we add a test for this change?

**Comment by shlomi-noach** [comment]
@harshit-gangal good idea; I don;t have an immediate understanding of how to test this, I'd need to dig in. If you happen to know, would you add such test?

File: go/vt/vttablet/tabletserver/query_executor.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/query_executor.go:668]: Can there be a future scenario where `NeedsReservedConn` is `true` even if the table is not `TEMPORARY`?

Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/query_executor.go]: I've meanwhile simplified the check for `IsTemporary()`

**Comment by shlomi-noach** [comment]
Fixes https://github.com/vitessio/vitess/issues/12146

**Comment by harshit-gangal** [comment]
> @harshit-gangal good idea; I don;t have an immediate understanding of how to test this, I'd need to dig in. If you happen to know, would you add such test?

I will add the test

File: go/vt/vttablet/tabletserver/query_executor.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vttablet/tabletserver/query_executor.go]: I do not expect that to change and I have added a test specifically for that. 
We can keep the current simplified version as well.

Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/query_executor.go]: let's keep the existing version, then, as it is more explicit :pray:

**Review by harshit-gangal — APPROVED** [review]

**Comment by shlomi-noach** [comment]
Tests-wise, I see validation that `NeedsReservedConn` is `true` or is `false`; but should there be a test that validates that `Reload` does, in fact, execute or does not execute?

**Review by frouioui — APPROVED** [review]

**Pull request merged**